slothy/targets/arm_v81m/arch_v81m.py

```diff
@@ -159,25 +158,72 @@ def extract(source, lbl):
             if state == 0:
                 p = loop_lbl_regexp.match(l_str)
                 if p is not None and p.group("label") == lbl:
-                    l.set_text(p.group("remainder"))
+                    l = l.copy().set_text(p.group("remainder"))
                     keep = True
                     state = 1
                 else:
                     pre.append(l)
                 continue
             if state == 1:
-                p = loop_end_regexp.match(l_str)
+                p = loop_end_regexp[loop_end_ctr].match(l_str)
                 if p is not None:
-                    state = 2
+                    if additional_data is None:
+                        additional_data = p.groupdict()
+                    loop_end_ctr += 1
+                    if loop_end_ctr == len(loop_end_regexp):
+                        state = 2
                     continue
                 body.append(l)
                 continue
             if state == 2:
                 post.append(l)
                 continue
         if state < 2:
-            raise Exception(f"Couldn't identify loop {lbl}")
-        return pre, body, post, lbl, ("lr", "lr", 0)
+            raise FatalParsingException(f"Couldn't identify loop {lbl}")
+        return pre, body, post, lbl, additional_data
+
+    @staticmethod 
+    def extract(source, lbl):
+        for loop_type in Loop.__subclasses__():
+            try:
+                l = loop_type(lbl)
+                return l._extract(source, lbl) + (l,)
```

Comment:
The `... + (l,)` may merit a comment. What's happening here? While you are at it, could you document `_extract()` the return tuple given by `_extract()` in a brief docstring?